### ISSUE TITLE
2.7.0: memory leak in TFLite's tflite::Interpreter::Invoke()

### ISSUE THREAD

**Issue by w3sip** [tensorflow/tensorflow#55503]
Seeing a memory leak in tflite::Interpreter::Invoke().
The leak is observed while running our software, wrapping TFLite 2.7.0 (built from source at that tag), on iOS 15.4 (EDIT: confirmed still leaking with 2.7.1 and 15.4.1), and using CoreML delegate.

The following leaks seem to occur roughly with every call to Invoke:

80 bytes chunk with the following stack:
```
class_createInstance		
__CFAllocateObject		
__NSSetI_new		
-[NSSet initWithArray:range:copyItems:]		
+[NSSet setWithArray:]		
-[MLDictionaryFeatureProvider featureNames]		
0x11e19b664		
0x11e14b44c		
tflite::Subgraph::Invoke()		
tflite::Interpreter::Invoke()	
```

48 bytes chunk with the following stack:
```
class_createInstance		
__CFAllocateObject		
__NSSetI_new		
-[NSSet initWithArray:range:copyItems:]		
+[NSSet setWithArray:]		
-[MLDictionaryFeatureProvider featureNames]		
0x11e19b664		
0x11e14b44c		
tflite::Subgraph::Invoke()		
tflite::Interpreter::Invoke()
```

16 bytes chunk with the following stack:
```
_CFCreateArrayStorage		
-[NSDictionary allKeys]		
-[MLDictionaryFeatureProvider featureNames]		
0x11e19b664		
0x11e14b44c		
tflite::Subgraph::Invoke()		
tflite::Interpreter::Invoke()	
```

32 bytes chunk with the following stack 
```
_objc_rootAllocWithZone		
objc_alloc_init		
-[NSTaggedPointerString UTF8String]		
0x11e19afd0		
-[MLNeuralNetworkEngine verifyInputs:error:]		
-[MLNeuralNetworkEngine evaluateInputs:options:error:]		
__62-[MLNeuralNetworkEngine predictionFromFeatures:options:error:]_block_invoke		
0x10342e7bc		
_dispatch_lane_barrier_sync_invoke_and_complete		
-[MLNeuralNetworkEngine predictionFromFeatures:options:error:]		
0x11e19b714		
0x11e14b44c		
tflite::Subgraph::Invoke()		
tflite::Interpreter::Invoke()	
```

Last time I went hunting for memory leaks, we were using 2.5.0, and there had been no leak there.

**Comment by mohantym** [comment]
Hi @w3sip ! Can you please share the steps to reproduce this behavior? Did you get a chance to check in TF 2.8 or nightly version too?

**Comment by w3sip** [comment]
- Unfortunately can't share the model, but it's a yolo v3 model.
- The leak is obvious, when running an inference under Instruments. It also seems to persist even after destroying the interpreter with TfLiteInterpreterDelete)
- Haven't checked in 2.8 (we'd have to upgrade CI to build it, hard dependency on Bazel makes things complex)
- did check in 2.6 (seems like the leak was already present there)

**Comment by mohantym** [comment]
Hi @sachinprasadhs ! Could you please look at this issue?

**Comment by w3sip** [comment]
One more note: the leak is gone if CoreML delegate is disabled.

**Comment by w3sip** [comment]
This seems relevant:
https://developer.apple.com/forums/thread/692425

**Comment by w3sip** [comment]
Looks like this had been addressed: https://github.com/tensorflow/tensorflow/commit/9f2e9d1e58f85a2b603baa1e682c0987fa49b203

Any chance this could be cherry-picked into a 2.7 update? In the meantime, we'll patch it locally and retest.

**Comment by sachinprasadhs** [comment]
Thanks for the issue report, I have created a PR for cherrypicking into r2.8 update.

**Comment by sachinprasadhs** [comment]
Hi, I see that you want this fix to be included in 2.7 version, generally cherrypicks will be done if it is security fix or a large community is effected with the bug. 
Since the fix is available in 2.9 and it's release is around the corner. You can use Tf-nightly for the immediate fix and then use the Tensorflow 2.9 when it is released. You can close this issue. Thanks!
